File: src/Blokemon.Product/Decks.cs
using System.Collections.Immutable;
using Blokemon.Core.SetDesign;

namespace Blokemon.Product;

public sealed record DeckRevision
{
    private DeckRevision(long value) => Value = value;

    public long Value { get; }

    public static DeckRevision Initial { get; } = new(1);

    public static DomainResult<DeckRevision, DeckRevisionFailure> Create(long value) =>
        value < 1
            ? DomainResult<DeckRevision, DeckRevisionFailure>.Failure(
                DeckRevisionFailure.MustBePositive
            )
            : DomainResult<DeckRevision, DeckRevisionFailure>.Success(new DeckRevision(value));

    internal bool TryNext(out DeckRevision next)
    {
        if (Value == long.MaxValue)
        {
            next = this;
            return false;
        }

        next = new DeckRevision(Value + 1);
        return true;
    }

    public override string ToString() =>
        Value.ToString(System.Globalization.CultureInfo.InvariantCulture);
}

public enum DeckRevisionFailure
{
    MustBePositive,
}

public sealed record DeckCardSelection(CardId CardId, int Quantity);

public abstract record DeckValidationIssue
{
    private DeckValidationIssue() { }

    public abstract TResult Match<TResult>(
        Func<CardId, int, TResult> onQuantityMustBePositive,
        Func<long, int, TResult> onWrongCardCount,
        Func<CardId, TResult> onUnknownCard,
        Func<CardId, long, int, TResult> onMechanicalCopyLimitExceeded,
        Func<TResult> onRegularCollectibleRequired,
        Func<CardId, long, int, TResult> onCollectibleQuantityNotOwned,
        Func<CardId, TResult> onCatalogueCardNotFree
    );

    public sealed record QuantityMustBePositive(CardId CardId, int Quantity) : DeckValidationIssue
    {
        public override TResult Match<TResult>(
            Func<CardId, int, TResult> onQuantityMustBePositive,
            Func<long, int, TResult> onWrongCardCount,
            Func<CardId, TResult> onUnknownCard,
            Func<CardId, long, int, TResult> onMechanicalCopyLimitExceeded,
            Func<TResult> onRegularCollectibleRequired,
            Func<CardId, long, int, TResult> onCollectibleQuantityNotOwned,
            Func<CardId, TResult> onCatalogueCardNotFree
        ) => onQuantityMustBePositive(CardId, Quantity);
    }

    public sealed record WrongCardCount(long Actual, int Required) : DeckValidationIssue
    {
        public override TResult Match<TResult>(
            Func<CardId, int, TResult> onQuantityMustBePositive,
            Func<long, int, TResult> onWrongCardCount,
            Func<CardId, TResult> onUnknownCard,
            Func<CardId, long, int, TResult> onMechanicalCopyLimitExceeded,
            Func<TResult> onRegularCollectibleRequired,
            Func<CardId, long, int, TResult> onCollectibleQuantityNotOwned,
            Func<CardId, TResult> onCatalogueCardNotFree
        ) => onWrongCardCount(Actual, Required);
    }

    public sealed record UnknownCard(CardId CardId) : DeckValidationIssue
    {
        public override TResult Match<TResult>(
            Func<CardId, int, TResult> onQuantityMustBePositive,
            Func<long, int, TResult> onWrongCardCount,
            Func<CardId, TResult> onUnknownCard,
            Func<CardId, long, int, TResult> onMechanicalCopyLimitExceeded,
            Func<TResult> onRegularCollectibleRequired,
            Func<CardId, long, int, TResult> onCollectibleQuantityNotOwned,
            Func<CardId, TResult> onCatalogueCardNotFree
        ) => onUnknownCard(CardId);
    }

    public sealed record MechanicalCopyLimitExceeded(CardId CardId, long Actual, int Allowed)
        : DeckValidationIssue
    {
        public override TResult Match<TResult>(
            Func<CardId, int, TResult> onQuantityMustBePositive,
            Func<long, int, TResult> onWrongCardCount,
            Func<CardId, TResult> onUnknownCard,
            Func<CardId, long, int, TResult> onMechanicalCopyLimitExceeded,
            Func<TResult> onRegularCollectibleRequired,
            Func<CardId, long, int, TResult> onCollectibleQuantityNotOwned,
            Func<CardId, TResult> onCatalogueCardNotFree
        ) => onMechanicalCopyLimitExceeded(CardId, Actual, Allowed);
    }

    public sealed record RegularCollectibleRequired : DeckValidationIssue
    {
        public override TResult Match<TResult>(
            Func<CardId, int, TResult> onQuantityMustBePositive,
            Func<long, int, TResult> onWrongCardCount,
            Func<CardId, TResult> onUnknownCard,
            Func<CardId, long, int, TResult> onMechanicalCopyLimitExceeded,
            Func<TResult> onRegularCollectibleRequired,
            Func<CardId, long, int, TResult> onCollectibleQuantityNotOwned,
            Func<CardId, TResult> onCatalogueCardNotFree
        ) => onRegularCollectibleRequired();
    }

    public sealed record CollectibleQuantityNotOwned(CardId CardId, long Requested, int Owned)
        : DeckValidationIssue
    {
        public override TResult Match<TResult>(
            Func<CardId, int, TResult> onQuantityMustBePositive,
            Func<long, int, TResult> onWrongCardCount,
            Func<CardId, TResult> onUnknownCard,
            Func<CardId, long, int, TResult> onMechanicalCopyLimitExceeded,
            Func<TResult> onRegularCollectibleRequired,
            Func<CardId, long, int, TResult> onCollectibleQuantityNotOwned,
            Func<CardId, TResult> onCatalogueCardNotFree
        ) => onCollectibleQuantityNotOwned(CardId, Requested, Owned);
    }

    public sealed record CatalogueCardNotFree(CardId CardId) : DeckValidationIssue
    {
        public override TResult Match<TResult>(
            Func<CardId, int, TResult> onQuantityMustBePositive,
            Func<long, int, TResult> onWrongCardCount,
            Func<CardId, TResult> onUnknownCard,
            Func<CardId, long, int, TResult> onMechanicalCopyLimitExceeded,
            Func<TResult> onRegularCollectibleRequired,
            Func<CardId, long, int, TResult> onCollectibleQuantityNotOwned,
            Func<CardId, TResult> onCatalogueCardNotFree
        ) => onCatalogueCardNotFree(CardId);
    }
}

public abstract record DeckValidationResult
{
    private DeckValidationResult() { }

    public bool IsValid => this is Valid;

    public abstract TResult Match<TResult>(
        Func<ValidatedDeck, TResult> onValid,
        Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalid
    );

    internal static DeckValidationResult Success(ValidatedDeck deck) => new Valid(deck);

    internal static DeckValidationResult Failure(ImmutableArray<DeckValidationIssue> issues) =>
        new Invalid(issues);

    public sealed record Valid(ValidatedDeck Deck) : DeckValidationResult
    {
        public override TResult Match<TResult>(
            Func<ValidatedDeck, TResult> onValid,
            Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalid
        ) => onValid(Deck);
    }

    public sealed record Invalid(ImmutableArray<DeckValidationIssue> Issues) : DeckValidationResult
    {
        public override TResult Match<TResult>(
            Func<ValidatedDeck, TResult> onValid,
            Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalid
        ) => onInvalid(Issues);
    }
}

public sealed class ValidatedDeck
{
    internal ValidatedDeck(ImmutableDictionary<CardId, int> cards) => Cards = cards;

    public ImmutableDictionary<CardId, int> Cards { get; }
}

public static class DeckValidator
{
    public const int RequiredCardCount = 60;

    public const int MechanicalCopyLimit = 4;

    public static DeckValidationResult Validate(
        LocalProfile profile,
        BlokemonRuntimeManifest authority,
        IEnumerable<DeckCardSelection> selections
    )
    {
        ArgumentNullException.ThrowIfNull(profile);
        ArgumentNullException.ThrowIfNull(authority);
        ArgumentNullException.ThrowIfNull(selections);

        var issues = ImmutableArray.CreateBuilder<DeckValidationIssue>();
        var quantities = new Dictionary<CardId, long>();

        foreach (var selection in selections)
        {
            ArgumentNullException.ThrowIfNull(selection);
            ArgumentNullException.ThrowIfNull(selection.CardId);

            if (selection.Quantity <= 0)
            {
                issues.Add(
                    new DeckValidationIssue.QuantityMustBePositive(
                        selection.CardId,
                        selection.Quantity
                    )
                );
                continue;
            }

            quantities[selection.CardId] =
                quantities.GetValueOrDefault(selection.CardId) + selection.Quantity;
        }

        var cardCount = quantities.Values.Sum();
        if (cardCount != RequiredCardCount)
        {
            issues.Add(new DeckValidationIssue.WrongCardCount(cardCount, RequiredCardCount));
        }

        var collectibles = authority.Collectibles.ToDictionary(
            static card => card.Id,
            StringComparer.Ordinal
        );
        var kits = authority.Kits.ToDictionary(static card => card.Id, StringComparer.Ordinal);
        var basicVim = authority.BasicVim.ToDictionary(
            static card => card.Id,
            StringComparer.Ordinal
        );
        var includesRegular = false;

        foreach (var (cardId, quantity) in quantities)
        {
            if (collectibles.TryGetValue(cardId.Value, out var collectible))
            {
                includesRegular |= collectible.Rank == BlokemonRank.Regular;
                CheckCopyLimit(cardId, quantity, collectible.StackCopyLimit, issues);

                var owned = profile.OwnedCollectibleQuantity(cardId);
                if (quantity > owned)
                {
                    issues.Add(
                        new DeckValidationIssue.CollectibleQuantityNotOwned(cardId, quantity, owned)
                    );
                }

                continue;
            }

            if (kits.TryGetValue(cardId.Value, out var kit))
            {
                if (!kit.FreelyAvailable)
                {
                    issues.Add(new DeckValidationIssue.CatalogueCardNotFree(cardId));
                }

                CheckCopyLimit(cardId, quantity, kit.StackCopyLimit, issues);
                continue;
            }

            if (basicVim.TryGetValue(cardId.Value, out var vim))
            {
                if (!vim.FreelyAvailable)
                {
                    issues.Add(new DeckValidationIssue.CatalogueCardNotFree(cardId));
                }

                continue;
            }

            issues.Add(new DeckValidationIssue.UnknownCard(cardId));
        }

        if (!includesRegular)
        {
            issues.Add(new DeckValidationIssue.RegularCollectibleRequired());
        }

        if (issues.Count > 0)
        {
            return DeckValidationResult.Failure(issues.ToImmutable());
        }

        var cards = quantities.ToImmutableDictionary(
            static entry => entry.Key,
            static entry => (int)entry.Value
        );
        return DeckValidationResult.Success(new ValidatedDeck(cards));
    }

    private static void CheckCopyLimit(
        CardId cardId,
        long quantity,
        int cardCopyLimit,
        ImmutableArray<DeckValidationIssue>.Builder issues
    )
    {
        var allowed = Math.Min(cardCopyLimit, MechanicalCopyLimit);
        if (quantity > allowed)
        {
            issues.Add(
                new DeckValidationIssue.MechanicalCopyLimitExceeded(cardId, quantity, allowed)
            );
        }
    }
}

public sealed class SavedDeck
{
    internal SavedDeck(
        DeckId id,
        DeckName name,
        DeckRevision revision,
        ImmutableDictionary<CardId, int> cards
    )
    {
        Id = id;
        Name = name;
        Revision = revision;
        Cards = cards;
    }

    public DeckId Id { get; }

    public DeckName Name { get; }

    public DeckRevision Revision { get; }

    public ImmutableDictionary<CardId, int> Cards { get; }
}

public abstract record DeckSaveFailure
{
    private DeckSaveFailure() { }

    public abstract TResult Match<TResult>(
        Func<DeckId, TResult> onAlreadyExists,
        Func<DeckId, TResult> onNotFound,
        Func<DeckId, DeckRevision, DeckRevision, TResult> onStaleRevision,
        Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalidDeck,
        Func<DeckId, TResult> onRevisionExhausted
    );

    public sealed record AlreadyExists(DeckId DeckId) : DeckSaveFailure
    {
        public override TResult Match<TResult>(
            Func<DeckId, TResult> onAlreadyExists,
            Func<DeckId, TResult> onNotFound,
            Func<DeckId, DeckRevision, DeckRevision, TResult> onStaleRevision,
            Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalidDeck,
            Func<DeckId, TResult> onRevisionExhausted
        ) => onAlreadyExists(DeckId);
    }

    public sealed record NotFound(DeckId DeckId) : DeckSaveFailure
    {
        public override TResult Match<TResult>(
            Func<DeckId, TResult> onAlreadyExists,
            Func<DeckId, TResult> onNotFound,
            Func<DeckId, DeckRevision, DeckRevision, TResult> onStaleRevision,
            Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalidDeck,
            Func<DeckId, TResult> onRevisionExhausted
        ) => onNotFound(DeckId);
    }

    public sealed record StaleRevision(
        DeckId DeckId,
        DeckRevision ExpectedRevision,
        DeckRevision ActualRevision
    ) : DeckSaveFailure
    {
        public override TResult Match<TResult>(
            Func<DeckId, TResult> onAlreadyExists,
            Func<DeckId, TResult> onNotFound,
            Func<DeckId, DeckRevision, DeckRevision, TResult> onStaleRevision,
            Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalidDeck,
            Func<DeckId, TResult> onRevisionExhausted
        ) => onStaleRevision(DeckId, ExpectedRevision, ActualRevision);
    }

    public sealed record InvalidDeck(ImmutableArray<DeckValidationIssue> Issues) : DeckSaveFailure
    {
        public override TResult Match<TResult>(
            Func<DeckId, TResult> onAlreadyExists,
            Func<DeckId, TResult> onNotFound,
            Func<DeckId, DeckRevision, DeckRevision, TResult> onStaleRevision,
            Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalidDeck,
            Func<DeckId, TResult> onRevisionExhausted
        ) => onInvalidDeck(Issues);
    }

    public sealed record RevisionExhausted(DeckId DeckId) : DeckSaveFailure
    {
        public override TResult Match<TResult>(
            Func<DeckId, TResult> onAlreadyExists,
            Func<DeckId, TResult> onNotFound,
            Func<DeckId, DeckRevision, DeckRevision, TResult> onStaleRevision,
            Func<ImmutableArray<DeckValidationIssue>, TResult> onInvalidDeck,
            Func<DeckId, TResult> onRevisionExhausted
        ) => onRevisionExhausted(DeckId);
    }
}

public sealed record DeckSaveTransition(LocalProfile Profile, SavedDeck Deck);

File: src/Blokemon.Product/Decks.fs
namespace Blokemon.Product

open System
open System.Collections.Immutable
open System.Globalization

/// Why a deck revision was rejected.
type DeckRevisionFailure =
    | MustBePositive = 0

/// A saved deck's optimistic-concurrency revision.
type DeckRevision =
    private
        { value: int64 }

    /// The revision every deck is first saved at.
    static member val Initial = { value = 1L }

    member this.Value = this.value

    override this.ToString() =
        this.value.ToString(CultureInfo.InvariantCulture)

    static member op_Equality(left: DeckRevision, right: DeckRevision) = left.Equals(right)

    static member op_Inequality(left: DeckRevision, right: DeckRevision) = not (left.Equals(right))

    static member Create(value: int64) =
        if value < 1L then
            DomainResult.Failed DeckRevisionFailure.MustBePositive
        else
            DomainResult.Succeeded { value = value }

    /// The next revision, or nothing when the counter has run out.
    member internal this.TryNext() =
        if this.value = Int64.MaxValue then
            None
        else
            Some { value = this.value + 1L }

/// How many copies of one card a deck asks for.
type DeckCardSelection =
    { CardId: CardId
      Quantity: int }

    static member op_Equality(left: DeckCardSelection, right: DeckCardSelection) =
        left.Equals(right)

    static member op_Inequality(left: DeckCardSelection, right: DeckCardSelection) =
        not (left.Equals(right))

/// Why a set of deck selections is not a legal deck.
[<RequireQualifiedAccess>]
type DeckValidationIssue =
    | QuantityMustBePositive of CardId: CardId * Quantity: int
    | WrongCardCount of Actual: int64 * Required: int
    | UnknownCard of CardId: CardId
    | MechanicalCopyLimitExceeded of CardId: CardId * Actual: int64 * Allowed: int
    | RegularCollectibleRequired
    | CollectibleQuantityNotOwned of CardId: CardId * Requested: int64 * Owned: int
    | CatalogueCardNotFree of CardId: CardId

    /// Folds the issue into a single value.
    member this.Match<'TResult>
        (
            onQuantityMustBePositive: Func<CardId, int, 'TResult>,
            onWrongCardCount: Func<int64, int, 'TResult>,
            onUnknownCard: Func<CardId, 'TResult>,
            onMechanicalCopyLimitExceeded: Func<CardId, int64, int, 'TResult>,
            onRegularCollectibleRequired: Func<'TResult>,
            onCollectibleQuantityNotOwned: Func<CardId, int64, int, 'TResult>,
            onCatalogueCardNotFree: Func<CardId, 'TResult>
        ) =
        match this with
        | DeckValidationIssue.QuantityMustBePositive(cardId, quantity) ->
            onQuantityMustBePositive.Invoke(cardId, quantity)
        | DeckValidationIssue.WrongCardCount(actual, required) ->
            onWrongCardCount.Invoke(actual, required)
        | DeckValidationIssue.UnknownCard cardId -> onUnknownCard.Invoke cardId
        | DeckValidationIssue.MechanicalCopyLimitExceeded(cardId, actual, allowed) ->
            onMechanicalCopyLimitExceeded.Invoke(cardId, actual, allowed)
        | DeckValidationIssue.RegularCollectibleRequired -> onRegularCollectibleRequired.Invoke()
        | DeckValidationIssue.CollectibleQuantityNotOwned(cardId, requested, owned) ->
            onCollectibleQuantityNotOwned.Invoke(cardId, requested, owned)
        | DeckValidationIssue.CatalogueCardNotFree cardId -> onCatalogueCardNotFree.Invoke cardId

/// Deck selections that passed every rule.
[<Sealed>]
type ValidatedDeck internal (cards: ImmutableDictionary<CardId, int>) =
    member _.Cards = cards

/// The outcome of validating deck selections.
[<RequireQualifiedAccess>]
type DeckValidationResult =
    | Valid of Deck: ValidatedDeck
    | Invalid of Issues: ImmutableArray<DeckValidationIssue>

/// A deck the player saved.
[<Sealed>]
type SavedDeck
    internal
    (id: DeckId, name: DeckName, revision: DeckRevision, cards: ImmutableDictionary<CardId, int>) =

    member _.Id = id

    member _.Name = name

    member _.Revision = revision

    member _.Cards = cards

/// Why a deck could not be saved.
[<RequireQualifiedAccess>]
type DeckSaveFailure =
    | AlreadyExists of DeckId: DeckId
    | NotFound of DeckId: DeckId
    | StaleRevision of
        DeckId: DeckId *
        ExpectedRevision: DeckRevision *
        ActualRevision: DeckRevision
    | InvalidDeck of Issues: ImmutableArray<DeckValidationIssue>
    | RevisionExhausted of DeckId: DeckId

    /// Folds the failure into a single value.
    member this.Match<'TResult>
        (
            onAlreadyExists: Func<DeckId, 'TResult>,
            onNotFound: Func<DeckId, 'TResult>,
            onStaleRevision: Func<DeckId, DeckRevision, DeckRevision, 'TResult>,
            onInvalidDeck: Func<ImmutableArray<DeckValidationIssue>, 'TResult>,
            onRevisionExhausted: Func<DeckId, 'TResult>
        ) =
        match this with
        | DeckSaveFailure.AlreadyExists deckId -> onAlreadyExists.Invoke deckId
        | DeckSaveFailure.NotFound deckId -> onNotFound.Invoke deckId
        | DeckSaveFailure.StaleRevision(deckId, expectedRevision, actualRevision) ->
            onStaleRevision.Invoke(deckId, expectedRevision, actualRevision)
        | DeckSaveFailure.InvalidDeck issues -> onInvalidDeck.Invoke issues
        | DeckSaveFailure.RevisionExhausted deckId -> onRevisionExhausted.Invoke deckId

/// Why a deck could not be deleted.
type DeckDeleteFailure =
    | NotFound = 0
